keps/sig-instrumentation/4828-component-flagz/README.md

```diff
@@ -0,0 +1,388 @@
+- Feature implemented for apiserver
+- Ensure proper tests are in place.
+
+#### Beta
+
+- Gather feedback from developers
+- Feature implemented for other Kubernetes Components
+
+#### GA
+
+- Several cycles of bake-time
+
+#### Deprecation
+
+We plan to use the existing strategy for deprecating the zpage as is used for metrics currently, that is
+
+- Announce deprecation and support policy of the z-page
+- Two versions passed since introducing the functionality that deprecates the z-page (to address version skew)
+- Address feedback on usage/changed behavior, provided on GitHub issues
+- Deprecate the z-page
+
+### Upgrade / Downgrade Strategy
+
+In alpha, no changes are required to maintain previous behavior. And the feature gate can be turned on to make use of the enhancement.
+
+### Version Skew Strategy
+
+The primary purpose of the flagz page is to provide information about the individual component itself. While the underlying data format may be enhanced over time, the endpoint itself will remain a reliable source of flag information across Kubernetes versions.
+
+## Production Readiness Review Questionnaire
+
+### Feature Enablement and Rollback
+
+We will target this feature behind a flag `ComponentFlagzFeatureGate`
+
+###### How can this feature be enabled / disabled in a live cluster?
+
+- [x] Feature gate (also fill in values in `kep.yaml`)
+  - Feature gate name: ComponentFlagzFeatureGate
+  - Components depending on the feature gate:
+    - apiserver
+    - kubelet
+    - scheduler
+    - controller-manager
+    - kube-proxy
+
+###### Does enabling the feature change any default behavior?
+
+Yes it will expose a new flagz endpoint. 
+
+###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?
+
+Yes. But it will remove the flagz endpoint
+
+###### What happens if we reenable the feature if it was previously rolled back?
+
+It will expose the flagz endpoint again
+
+###### Are there any tests for feature enablement/disablement?
+
+Unit test will be introduced in alpha implementation.
+
+### Rollout, Upgrade and Rollback Planning
+
+<!--
+This section must be completed when targeting beta to a release.
+-->
+
+###### How can a rollout or rollback fail? Can it impact already running workloads?
+
+This feature should not cause rollout failures. If it does, we can disable the feature. In the worst
+case, it is possible it could cause runtime failures, but it is highly unlikely we would not detect this
+with existing tests.
+
+###### What specific metrics should inform a rollback?
+
+<!--
+What signals should users be paying attention to when the feature is young
+that might indicate a serious problem?
+-->
+
+- apiserver_request_duration_seconds metric that will tell us if there's a spike in request latency of kube-apiserver that might indicate that the flagz endpoint is interfering with the component's core functionality or causing delays in processing other requests
```

Comment:
IIRC you won't get this metric for free, you'll have to make sure that the handler for the flagz endpoint is instrumented.

Reply:
Umm unsure about how flagz handler relates with this metric which I thought is enabled by default? The idea was to use this metric to monitor general load of requests that apiserver handles and use that as a signal to check if flagz is consuming all system resources causing delays in apiserver's capability to serve other resource requests.. 

though yeah I guess that wouldnt be a clean signal to detect issues with the flagz endpoint in a deterministic way.. Do you suggest introducing a new metric that keeps track of request latency for just /flagz requests ?

Reply:
I forgot to submit my comment on your PR, but you already dealt with that concern of mine in your POC: https://github.com/kubernetes/kubernetes/pull/127581#discussion_r1775901244